### PR TITLE
[SC-1037] VelodromeV2

### PR DESCRIPTION
**Pull request by zZoMROT**
To do before merging:
- [x] add `secrets.OPTIMISTIC_RPC_URL`
- [x] change `secrets.MAINNET_RPC_URL` to `secrets.MAINNET_RPC_URL|secrets.RPC_AUTH_HEADER`
- [x] resolve conversations

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -20,8 +20,8 @@ jobs:
       - uses: ./.github/actions/setup
       - run: yarn test:ci
         env:
-          MAINNET_RPC_URL: ${{ secrets.MAINNET_RPC_URL }}
-          RPC_AUTH_HEADER: ${{ secrets.RPC_AUTH_HEADER }}
+          MAINNET_RPC_URL: "${{ secrets.MAINNET_RPC_URL }}|${{ secrets.RPC_AUTH_HEADER }}"
```
Comment (zZoMROT) [inline .github/workflows/test.yml:23]:
```suggestion
          MAINNET_RPC_URL: ${{ secrets.MAINNET_RPC_URL }}
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -30,8 +30,8 @@ jobs:
       - uses: ./.github/actions/setup
       - run: yarn coverage
         env:
-          MAINNET_RPC_URL: ${{ secrets.MAINNET_RPC_URL }}
-          RPC_AUTH_HEADER: ${{ secrets.RPC_AUTH_HEADER }}
+          MAINNET_RPC_URL: "${{ secrets.MAINNET_RPC_URL }}|${{ secrets.RPC_AUTH_HEADER }}"
```
Comment (zZoMROT) [inline .github/workflows/test.yml:33]:
```suggestion
          MAINNET_RPC_URL: ${{ secrets.MAINNET_RPC_URL }}
```

File: test/oracles/VelodromeV2Oracle.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
+const { loadFixture } = require('@nomicfoundation/hardhat-network-helpers');
+const { network } = require('hardhat');
+const { assertRoughlyEqualValues, deployContract } = require('@1inch/solidity-utils');
+const {
+    tokens,
+    deployParams: { VelodromeV2, UniswapV3 },
+    defaultValues: { thresholdFilter },
+} = require('../helpers.js');
+
+describe.skip('VelodromeV2Oracle', function () {
```
Comment (zZoMROT) [inline test/oracles/VelodromeV2Oracle.js:10]:
```suggestion
describe('VelodromeV2Oracle', function () {
```

**Review by ZumZoom — APPROVED** [review]

File: contracts/oracles/VelodromeV2Oracle.sol [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    IVelodramoV2Router public immutable ROUTER;
+    IVelodramoV2Registry public immutable REGISTRY;
+
+    constructor(IVelodramoV2Router _router, IVelodramoV2Registry _registry) {
+        ROUTER = _router;
+        REGISTRY = _registry;
+    }
+
+    function getRate(IERC20 srcToken, IERC20 dstToken, IERC20 connector, uint256 thresholdFilter) external view override returns (uint256 rate, uint256 weight) {
```
Comment (byshape) [inline contracts/oracles/VelodromeV2Oracle.sol:34]: Did you intentionally choose not to write natspec docs for some reason?

File: contracts/oracles/VelodromeV2Oracle.sol [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        address[] memory factories = REGISTRY.poolFactories();
+        uint256 factoriesLength = factories.length;
+        OraclePrices.Data memory ratesAndWeights = OraclePrices.init(2 * factoriesLength);
+        uint256 b0;
+        uint256 b1;
+        if (connector == _NONE) {
+            for (uint256 i = 0; i < factoriesLength; i++) {
```
Comment (byshape) [inline contracts/oracles/VelodromeV2Oracle.sol:41]: It seems that the following code can be separated into a new function that will be called 3 times:
```solidity
for (uint256 i = 0; i < factoriesLength; i++) {
    (b0, b1) = _getReserves(srcToken, dstToken, true, factories[i]);
    if (b0 > 0) {
        ratesAndWeights.append(OraclePrices.OraclePrice(Math.mulDiv(b1, 1e18, b0), (b0 * b1).sqrt()));
    }
    (b0, b1) = _getReserves(srcToken, dstToken, false, factories[i]);
    if (b0 > 0) {
        ratesAndWeights.append(OraclePrices.OraclePrice(Math.mulDiv(b1, 1e18, b0), (b0 * b1).sqrt()));
    }
}
(rate, weight) = ratesAndWeights.getRateAndWeight(thresholdFilter);
```

**Review by byshape — CHANGES_REQUESTED** [review]
It will be great to discuss the possibility of separating duplicated code into a new function and its impact on the overall code readability.

File: contracts/oracles/VelodromeV2Oracle.sol (hunk 4 above)
Comment (zZoMROT) [reply contracts/oracles/VelodromeV2Oracle.sol:34]: That's a good observation. It seems like a useful idea to add NatSpec comments, at least in the interface, to enhance documentation and understanding

**Pull request merged**